core/interop/src/loaders/predicate.rs

```diff
@@ -0,0 +1,183 @@
+    Inner: Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("PredicateModuleLoader")
+            .field("predicate", &"Fn()")
+            .field("inner", &self.inner)
+            .finish()
+    }
+}
+
+impl<P, Inner> PredicateModuleLoader<P, Inner>
+where
+    P: Fn(Option<&Path>, JsString) -> JsResult<JsString>,
+{
+    /// Create a new module loader.
+    pub fn new(predicate: P, inner: Inner) -> Self {
+        Self { predicate, inner }
+    }
+}
+
+impl<P, Inner> ModuleLoader for PredicateModuleLoader<P, Inner>
+where
+    P: Fn(Option<&Path>, JsString) -> JsResult<JsString>,
+    Inner: ModuleLoader,
+{
+    fn load_imported_module(
+        &self,
+        referrer: Referrer,
+        specifier: JsString,
+        finish_load: Box<dyn FnOnce(JsResult<Module>, &mut Context)>,
+        context: &mut Context,
+    ) {
+        match (self.predicate)(referrer.path(), specifier) {
+            Ok(specifier) => {
+                self.inner
+                    .load_imported_module(referrer, specifier, finish_load, context);
+            }
+            Err(err) => finish_load(Err(err), context),
+        }
+    }
+}
+
+/// A collection of predicate functions to use with the [`PredicateModuleLoader`].
+pub mod predicates {
+    use std::path::{Component, Path, PathBuf};
+
+    use boa_engine::{js_string, JsError, JsResult, JsString};
+
+    /// A predicate that only allows loading absolute modules. Will refuse to load
+    /// any specifier that starts with `./` or `../`.
+    ///
+    /// # Errors
+    /// This predicate will return an error if the specifier is a relative path (starts with
+    /// `./` or `../`).
+    #[inline]
+    #[allow(clippy::needless_pass_by_value)]
+    pub fn absolute_only(_: Option<&Path>, specifier: JsString) -> JsResult<JsString> {
+        let specifier = specifier.to_std_string_escaped();
+        let short_path = Path::new(&specifier);
+
+        if short_path.starts_with(".") || short_path.starts_with("..") {
+            Err(JsError::from_opaque(js_string!("relative path").into()))
+        } else {
+            Ok(JsString::from(specifier))
+        }
+    }
+
+    /// A Predicate that can be used with the `PrediceModuleLoader` that resolves paths
+    /// from the referrer and the specifier, normalize the paths and ensure the path
+    /// is within a base.
+    ///
+    /// If the base is empty, that last verification will be skipped. This can be used
+    /// for module loaders that don't have an actual filesystem as their backend
+    /// (such as [`crate::loaders::HashMapModuleLoader`]).
+    ///
+    /// # Errors
+    /// This predicate will return an error if the specifier is relative but the referrer
+    /// does not have a path, or if the resolved path is outside `base`.
+    #[inline]
+    pub fn path_resolver(base: PathBuf) -> impl Fn(Option<&Path>, JsString) -> JsResult<JsString> {
```

Comment:
@jedel1043 This function will fix the bug in the `SimpleModuleLoader`. I will move this code into the `SimpleModuleLoader` as a free function, and use that function here instead. It should make it easier.

I'll duplicate the tests as well.